src/simd/runtime.rs

```diff
@@ -0,0 +1,53 @@
+use crate::iter::Bytes;
+use super::avx2;
+use super::sse42;
+
+pub enum RuntimeFeature {
+    Avx2,
+    Sse42,
+    Nop,
+}
+
+impl RuntimeFeature {
+    fn detect() -> Self {
+        if is_x86_feature_detected!("avx2") {
+            RuntimeFeature::Avx2
+        } else if is_x86_feature_detected!("sse4.2") {
+            RuntimeFeature::Sse42
+        } else {
+            RuntimeFeature::Nop
+        }
+    }
+}
+
+static mut RUNTIME_FEATURE: Option<RuntimeFeature> = None;
+
+pub fn match_uri_vectored(bytes: &mut Bytes) {
+    // SAFETY: calls are guarded by a feature check
+    unsafe {
+        if let None = RUNTIME_FEATURE {
+            RUNTIME_FEATURE = Some(RuntimeFeature::detect());
+        }
```

Comment:
This kind of unsynchronized access of a static could cause a data race if the code is called at the same time in separate threads.

Reply:
To my understanding this is correct, no thread/core should will perform a torn read or write given that we're storing/loading naturally aligned values of less than a quadword. [Ref AMD64 manual (7.3.2)](https://www.amd.com/system/files/TechDocs/24593.pdf#page=260):

> Cacheable, naturally-aligned single loads or stores of up to a quadword are atomic on any processor
model, as are misaligned loads or stores of less than a quadword that are contained entirely within a
naturally-aligned quadword.

Obviously there are other issues with shared globals, but in this instance it's a simple init once, worst case multiple threads will write, but it's idempotent so there's no observable issue.

The previous implementation which stored a shared pointer should also have been correct as long as the global is aligned. In this instance both `RuntimeFeature` and `Option<RuntimeFeature>` are a single byte which is always contained entirely with a naturally-aligned quadword.

These assumptions hold true:
```rs
#[test]
fn test_assumptions() {
    assert_eq!(std::mem::size_of::<RuntimeFeature>(), 1);
    assert_eq!(std::mem::size_of::<Option<RuntimeFeature>>(), 1);
    assert_eq!(RuntimeFeature::Avx2 as u8, 0);
    assert_eq!(RuntimeFeature::Sse42 as u8, 1);
    assert_eq!(RuntimeFeature::Nop as u8, 2);
    assert_eq!(unsafe { std::mem::transmute::<Option::<RuntimeFeature>, u8>(None) }, 3);
}
```

But I agree it's probably more idiomatic and obviously correct to wrap this in an atomic with a relaxed store which I believe will be codegen'd to single `mov`s anyway for the above reasons

Reply:
Made the change

Reply:
You're right. Sorry, I should have said more in my initial comment. On most architectures, it would be a non-tearing read/write. I had just checked in the Rust Nomicon, to see about the language semantics. I wanted to know, can we rely on that, or is there a theoretical instance it would be wrong. It makes no such promise, just claims these this is "unsound":

> - two or more threads concurrently accessing a location of memory
> - one or more of them is a write
> - one or more of them is unsynchronized